### PR TITLE
Enable Qthreads tasking shim in runtime

### PR DESCRIPTION
**Pull request by gbtitus**
With this change we complete the work necessary to bring the Qthreads tasking shim into the Chapel runtime.  Mostly this involved rewriting or working around the places where the shim called internal (non-API) Qthreads functions.

### CODE REVIEW

**Comment by gbtitus** [comment]
@ronawho , would you please review this one?

It's passed full testing with comm=none on linux64, comm=gasnet testing with multilocale and examples on linux64, comm=gasnet testing with examples on XC, and comm=ugni testing with example on XC using both the gcc and intel compilers.

File: third-party/qthread/Makefile [hunk 1]
```diff
@@ -115,13 +109,47 @@ qthread-config: FORCE
...
+# The if-checks on the scheduler value look the same here, but they're
+# actually asking different questions.  In the first, it's "Will remote
+# caching work with this Qthreads build?", which is true iff qthreads do
+# not move from one worker to another (thus invalidating TLS).  In the
+# second, it's "Is there only one worker per shepherd?", which changes
+# how the shim sets certain QT_* environment variables to parameterize
+# Qthreads behavior.  These have the same answer now, but in a future
+# scheduler they could differ.
+#
+ifeq (,$(findstring $(SCHEDULER),lifo mtsfifo mutexfifo nemesis))
```
Comment (ronawho) [inline third-party/qthread/Makefile]: Oh, one more thing. For both of these I think we should have an `ifeq (,$(findstring $(SCHEDULER),sherwood nottingham))` and have the else give an error

**Comment by ronawho** [comment]
Thanks for breaking this up into logical commits!

Overall, I think this looks good. The `__sync_synchronize` issue needs to be dealt with before merging since it will break the pgi build. I think the SPINLOCK body one is fine as is since it's not a performance critical section

File: third-party/qthread/Makefile (hunk 1 above)
Comment (gbtitus) [inline third-party/qthread/Makefile]: Yeah, I got lazy -- will do.

**Comment by gbtitus** [comment]
@ronawho , take a look at the updates?  Also, I tested the hellos with comm=none using the PGI compiler on XC, which I think should cover the whitebox case.

**Comment by ronawho** [comment]
Looks great to me!

Sidenote: it'd be amazing if our smoke test would run all the compilers instead of just intel. Might bug tony about that, because it's a pretty big pain to test all the backend compilers

**Comment by gbtitus** [comment]
I think I'll make one more commit.  The .h file is a mess; it contains a bunch of old/unnecessary stuff from its fifo origins.  I'm going to clean it up a bit.
